SyncClient refactoring

danbev opened the pull request.

danbev commented:
@secondsun Would you mind reviewing and let me know if there is anything missing.

danbev commented:
@secondsun Did you get chance to try this out? Let me know if you need any changes and we'll get them in.

secondsun commented on client/client-netty/src/main/java/org/jboss/aerogear/sync/client/netty/NettySyncClient.java:
```diff
@@ -93,109 +98,132 @@ protected void initChannel(SocketChannel ch) {
+    }
+
+    @Override
+    public String clientId() {
```
Isn't client ID supposed to be unique per client? (IE the tuple which defines a connection between client and server)

danbev replied:
We can make it a ```UUID``` if that helps? This is not used with the NettyClient so you let me know what would be best?  This would be something different from the clientId that belongs to the ClientDocument.

secondsun commented on client/client-netty/src/main/java/org/jboss/aerogear/sync/client/netty/NettySyncClient.java:
```diff
@@ -34,22 +34,26 @@
  * @param <T> The type of the Document that this client can handle
  * @param <S> The type of {@link Edit}s that this client can handle
  */
-public final class SyncClient<T, S extends Edit<? extends Diff>> extends Observable {
```
It will still be nice if the core sync stuff was an abstract class I can extend to use Android specific things.  Logging jumps to mind but also using broadcast receivers and friends instead of Observers is also a platform pattern.

danbev replied:
Ah I see. Let me take a closer look at this tomorrow.

danbev commented:
@secondsun I've updated with https://github.com/danbev/aerogear-sync-server/commit/e10f05dc3a3a57fae4727dff9994ab7d9cde1e60 and made some changes to the SyncClient and SyncClientEngine which I hope will address usage of Observable. 

I've updated my [demo](https://github.com/danbev/android-diffsync-demo/tree/sync-client-refactoring) to try this out but I'm not exactly sure how the broadcast stuff works and if it would fit nicely into this refactoring. The idea is that you could provide an implementation of [PatchObservable](https://github.com/danbev/aerogear-sync-server/blob/syncclient-refactoring/client/client-core/src/main/java/org/jboss/aerogear/sync/client/PatchObservable.java) and it would use Android Broadcast. This would then used when creating a [ClientSyncEngine](https://github.com/danbev/android-diffsync-demo/blob/sync-client-refactoring/diffsync-demo/src/main/java/org/aerogear/diffsync/android/demo/DiffSyncMainActivity.java#L76). 

I've also switched the logging to use Netty's internal logging for the Netty specific parts to avoid that dependency.

secondsun commented on client/client-netty/src/main/java/org/jboss/aerogear/sync/client/netty/NettySyncClient.java:
```diff
@@ -60,19 +64,20 @@
-        host = builder.host;
+    private NettySyncClient(final Builder<T, S> builder) {
+        host = checkNotNull(builder.host, "host must not be null");
+        path = checkNotNull(builder.path, "path must not be null");
```
What is path?

secondsun commented:
W00t!  It does the needful!

+1

danielpassos commented:
:+1:  Tested using https://github.com/secondsun/aerogear-android-cookbook/tree/SyncDemo/SyncDemo

danielpassos commented:
@danbev Could you rebase it?

danbev commented:
@danielpassos I've rebased it now. Will try it out with the [SyncDemo](https://github.com/secondsun/aerogear-android-cookbook/tree/SyncDemo/SyncDemo).

danbev commented:
@secondsun @danielpassos I've rebase once more as I've moved the client related interfaces from the _client-engine_ module to the _api_ module which makes more sense. 

I was able to build the SyncDemo and deploy it but I'm not sure what kind of Keycloak set up I'm expected to have running. Could you provide me with some instructions on what is needed for running the demo?

danielpassos commented:
@danbev You can use this branch => https://github.com/danielpassos/aerogear-android-cookbook/tree/SyncDemo/SyncDemo

danielpassos commented:
:+1:  Just again after the last change and still working like a charme

danbev commented:
@danielpassos Perfect, tried it and work great!

danielpassos commented:
Landed at https://github.com/aerogear/aerogear-sync-server/commit/bef27bbce0d69ecf3947d6f75e4387089c2aa88c
